File: csharp/Betfair.ESAClient/Betfair.ESAClient/Protocol/SubscriptionHandler.cs
using Betfair.ESASwagger.Model;
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Betfair.ESAClient.Protocol
{
    /// <summary>
    /// Generic subscription handler for change messages:
    /// 1) Tracks clocks to facilitate resubscripiton
    /// 2) Provides useful timings for initial image
    /// 3) Supports the ability to re-combine segmented messages to retain event level atomicity
    /// </summary>
    /// <typeparam name="S">Subscription request message type</typeparam>
    /// <typeparam name="C">Change message type</typeparam>
    /// <typeparam name="I">Change message item type</typeparam>
    public class SubscriptionHandler<S,C,I> where C : ChangeMessage<I> where S : RequestMessage
    {
        private int _subscriptionId;
        private bool _isSubscribed;
        private bool _isMergeSegments;
        private List<I> _mergedChanges;
        private Stopwatch _ttfm;
        private Stopwatch _ttlm;
        private int _itemCount;
        private TaskCompletionSource<bool> _subscriptionComplete = new TaskCompletionSource<bool>();

        public SubscriptionHandler(int id, S subscriptionMessage, bool isMergeSegments)
        {
            SubscriptionMessage = subscriptionMessage;
            _isMergeSegments = isMergeSegments;
            IsSubscribed = false;
            _subscriptionId = id;
            _ttfm = Stopwatch.StartNew();
            _ttlm = Stopwatch.StartNew();
        }

        internal void Cancel()
        {
            //unwind waiters
            _subscriptionComplete.TrySetCanceled();
        }


        public string InitialClk { get; private set; }
        public string Clk { get; private set; }
        public S SubscriptionMessage { get; private set; }
        public bool IsSubscribed { get; private set; }
        public long? HeartbeatMs { get; private set; }
        public long? ConflationMs { get; private set; }
        public long? LastPt { get; private set; }
        public DateTime LastArrivalTime { get; private set; }
        public TimeSpan HeartbeatInterval { get; private set; }

        public Task<bool> SubscriptionComplete
        {
            get
            {
                return _subscriptionComplete.Task;
            }
        }


        public C ProcessChangeMessage(C changeMessage)
        {
            if(_subscriptionId != changeMessage.Id)
            {
                //previous subscription id - ignore
                return null;
            }

            //Every message store timings
            LastPt = changeMessage.Pt;
            LastArrivalTime = changeMessage.ArrivalTime;

            if (changeMessage.IsStartOfRecovery)
            {
                //Start of recovery
                _ttfm.Stop();
                Trace.TraceInformation("{0}: Start of image", typeof(S).Name);
            }

            if (changeMessage.ChangeType == ChangeType.HEARTBEAT)
            {
                //Swallow heartbeats
                changeMessage = null;
            }
            else if(changeMessage.SegmentType != SegmentType.NONE && _isMergeSegments)
            {
                //Segmented message and we're instructed to merge (which makes segments look atomic).
                changeMessage = MergeMessage(changeMessage);
            }

            if(changeMessage != null)
            {
                //store clocks
                if(changeMessage.InitialClk != null)
                {
                    InitialClk = changeMessage.InitialClk;
                }
                if(changeMessage.Clk != null)
                {
                    Clk = changeMessage.Clk;
                }

                if (!_isSubscribed)
                {
                    //During recovery
                    if (changeMessage.Items != null)
                    {
                        _itemCount += changeMessage.Items.Count;
                    }
                }
                
                if (changeMessage.IsEndOfRecovery)
                {
                    //End of recovery
                    _isSubscribed = true;
                    HeartbeatMs = changeMessage.HeartbeatMs;
                    HeartbeatInterval = TimeSpan.FromMilliseconds((double)HeartbeatMs);
                    ConflationMs = changeMessage.ConflateMs;
                    _ttlm.Stop();
                    Trace.TraceInformation("{0}: End of image: type:{6}, ttfm:{1}, ttlm:{2}, conflation:{3}, heartbeat:{4}, change.items:{5}", 
                        typeof(S).Name, 
                        _ttfm.Elapsed, 
                        _ttlm.Elapsed, 
                        ConflationMs, 
                        HeartbeatMs,
                        _itemCount,
                        changeMessage.ChangeType);

                    //unwind future
                    _subscriptionComplete.TrySetResult(true);
                }
               
            }
            return changeMessage;
        }

        private C MergeMessage(C changeMessage)
        {
            //merge segmented messages so client sees atomic view across segments
            if (changeMessage.SegmentType == SegmentType.SEG_START)
            {
                //start merging
                _mergedChanges = new List<I>();
            }
            //accumulate
            _mergedChanges.AddRange(changeMessage.Items);

            if (changeMessage.SegmentType == SegmentType.SEG_END)
            {
                //finish merging
                changeMessage.SegmentType = SegmentType.NONE;
                changeMessage.Items = _mergedChanges;
                _mergedChanges = null;
            }
            else
            {
                //swallow message as we're still merging
                changeMessage = null;
            }
            return changeMessage;
        }
    }
}

File: java/client/src/main/java/com/betfair/esa/client/protocol/SubscriptionHandler.java
package com.betfair.esa.client.protocol;

import com.betfair.esa.swagger.model.RequestMessage;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.CountDownLatch;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.slf4j.profiler.StopWatch;

/**
 * Generic subscription handler for change messages: 1) Tracks clocks to facilitate resubscripiton
 * 2) Provides useful timings for initial image 3) Supports the ability to re-combine segmented
 * messages to retain event level atomicity Created by mulveyj on 07/07/2016.
 */
public class SubscriptionHandler<S extends RequestMessage, C extends ChangeMessage<I>, I> {
    private static final Logger LOG = LoggerFactory.getLogger(SubscriptionHandler.class);
    private final int subscriptionId;
    private final S subscriptionMessage;
    private boolean isSubscribed;
    private final boolean isMergeSegments;
    private List<I> mergedChanges;
    private final StopWatch ttfm;
    private final StopWatch ttlm;
    private int itemCount;
    private final CountDownLatch subscriptionComplete = new CountDownLatch(1);

    private long lastPublishTime;
    private long lastArrivalTime;
    private String initialClk;
    private String clk;
    private Long heartbeatMs;
    private Long conflationMs;

    public SubscriptionHandler(S subscriptionMessage, boolean isMergeSegments) {
        this.subscriptionMessage = subscriptionMessage;
        this.isMergeSegments = isMergeSegments;
        isSubscribed = false;
        subscriptionId = subscriptionMessage.getId();
        ttfm = new StopWatch("ttfm");
        ttlm = new StopWatch("ttlm");
    }

    public S getSubscriptionMessage() {
        return subscriptionMessage;
    }

    public boolean isSubscribed() {
        return isSubscribed;
    }

    public long getLastPublishTime() {
        return lastPublishTime;
    }

    public long getLastArrivalTime() {
        return lastArrivalTime;
    }

    public Long getHeartbeatMs() {
        return heartbeatMs;
    }

    public Long getConflationMs() {
        return conflationMs;
    }

    public String getInitialClk() {
        return initialClk;
    }

    public String getClk() {
        return clk;
    }

    void cancel() {
        // unwind waiters
        subscriptionComplete.countDown();
    }

    public C processChangeMessage(C changeMessage) {
        if (subscriptionId != changeMessage.getId()) {
            // previous subscription id - ignore
            return null;
        }

        // Every message store timings
        lastPublishTime = changeMessage.getPublishTime();
        lastArrivalTime = changeMessage.getArrivalTime();

        if (changeMessage.isStartOfRecovery()) {
            // Start of recovery
            ttfm.stop();
            LOG.info("{}: Start of image", subscriptionMessage.getOp());
        }

        if (changeMessage.getChangeType() == ChangeType.HEARTBEAT) {
            // Swallow heartbeats
            changeMessage = null;
        } else if (changeMessage.getSegmentType() != SegmentType.NONE && isMergeSegments) {
            // Segmented message and we're instructed to merge (which makes segments look atomic).
            changeMessage = MergeMessage(changeMessage);
        }

        if (changeMessage != null) {
            // store clocks
            if (changeMessage.getInitialClk() != null) {
                initialClk = changeMessage.getInitialClk();
            }
            if (changeMessage.getClk() != null) {
                clk = changeMessage.getClk();
            }

            if (!isSubscribed) {
                // During recovery
                if (changeMessage.getItems() != null) {
                    itemCount += changeMessage.getItems().size();
                }
            }

            if (changeMessage.isEndOfRecovery()) {
                // End of recovery
                isSubscribed = true;
                heartbeatMs = changeMessage.getHeartbeatMs();
                conflationMs = changeMessage.getConflateMs();
                ttlm.stop();
                LOG.info(
                        "{}: End of image: type:{}, ttfm:{}, ttlm:{}, conflation:{}, heartbeat:{},"
                                + " change.items:{}",
                        subscriptionMessage.getOp(),
                        changeMessage.getChangeType(),
                        ttfm,
                        ttlm,
                        conflationMs,
                        heartbeatMs,
                        itemCount);

                // unwind future
                subscriptionComplete.countDown();
            }
        }
        return changeMessage;
    }

    private C MergeMessage(C changeMessage) {
        // merge segmented messages so client sees atomic view across segments
        if (changeMessage.getSegmentType() == SegmentType.SEG_START) {
            // start merging
            mergedChanges = new ArrayList<>();
        }
        // accumulate
        mergedChanges.addAll(changeMessage.getItems());

        if (changeMessage.getSegmentType() == SegmentType.SEG_END) {
            // finish merging
            changeMessage.setSegmentType(SegmentType.NONE);
            changeMessage.setItems(mergedChanges);
            mergedChanges = null;
        } else {
            // swallow message as we're still merging
            changeMessage = null;
        }
        return changeMessage;
    }
}
